Lib/test/test_shelve.py

```diff
@@ -165,6 +168,277 @@ def test_default_protocol(self):
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(
+                            value, f"{len(type(array_data).__name__)}"
+                        )
+
+                        key, value = s.next()
+                        self.assertEqual("bytearray_data", key)
+                        self.assertEqual(
+                            value, f"{len(type(bytearray_data).__name__)}"
+                        )
+
+                        key, value = s.next()
+                        self.assertEqual("bytes_data", key)
+                        self.assertEqual(
+                            value, f"{len(type(bytes_data).__name__)}"
+                        )
+
+                        key, value = s.first()
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(
+                            value, f"{len(type(array_data).__name__)}"
+                        )
+                    else:
+                        key, value = s.set_location(b"foo")
+                        self.assertEqual("foo", key)
+                        self.assertEqual(value, "str")
+
+                        key, value = s.previous()
+                        self.assertEqual("bytes_data", key)
+                        self.assertEqual(value, "bytes")
+
+                        key, value = s.previous()
+                        self.assertEqual("bytearray_data", key)
+                        self.assertEqual(value, "bytearray")
+
+                        key, value = s.previous()
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(value, "array")
+
+                        key, value = s.next()
+                        self.assertEqual("bytearray_data", key)
+                        self.assertEqual(value, "bytearray")
+
+                        key, value = s.next()
+                        self.assertEqual("bytes_data", key)
+                        self.assertEqual(value, "bytes")
+
+                        key, value = s.first()
+                        self.assertEqual("array_data", key)
+                        self.assertEqual(value, "array")
+
+                        self.assertEqual(s["foo"], "str")
+                        self.assertEqual(s["bytes_data"], "bytes")
+                        self.assertEqual(s["bytearray_data"], "bytearray")
+                        self.assertEqual(s["array_data"], "array")
+
+        with self.assertRaises(AssertionError):
+            def serializer(obj, protocol=None):
+                return bytes(type(obj).__name__, 'utf-8')
+
+            def deserializer(data):
+                pass
+
+            with shelve.BsdDbShelf(berkeleydb.btopen(self.fn),
+                                   serializer=serializer,
+                                   deserializer=deserializer) as s:
+                s["foo"] = "bar"
+                self.assertEqual(s["foo"], "bar")
```

Comment:
I don't understand why you would expect this. What are you trying to test here actually?

Reply:
The purpose is to test an incomplete 'deserializer'.